### ISSUE TITLE
Use of reabstraction of default interface methods causes System.TypeLoadException: VTable setup of type failed

### ISSUE THREAD

**Issue by AlekseyTs** [mono/mono#14495]
It is possible that support for reabstraction was not implemented in Mono at all. Here is the link to relevant PR for CoreClr - https://github.com/dotnet/coreclr/pull/23313. 

Here is an example of the scenario. An attempt to compile and run the following code:
```
public interface I1
{
    void M1();
}

public interface I2 : I1
{
    abstract void I1.M1();
}

class Test1 : I2
{
    static void Main()
    {
        I1 i1 = new Test1();
        i1.M1();
    }

    void I1.M1()
    {
        System.Console.WriteLine(""Test1.M1"");
    }
}
```

causes an exception:
```
Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.DefaultInterfaceImplementationTests.MethodReAbstraction_03 [FAIL]
      Roslyn.Test.Utilities.ExecutionException : 
      Execution failed for assembly ''.
      Exception: System.TypeLoadException: VTable setup of type Test1 failed
        at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
        at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <04789743042047e7ac7c5211841abf13>:0 
      ---- System.TypeLoadException : VTable setup of type Test1 failed
      Stack Trace:
          at Roslyn.Test.Utilities.Desktop.DesktopRuntimeEnvironment.Execute (System.String moduleName, System.String[] args, System.String expectedOutput) [0x000ba] in <2741b86d93474cfcb49ec1def5701445>:0 
          at Microsoft.CodeAnalysis.Test.Utilities.CompilationVerifier.Emit (System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, System.Collections.Generic.IEnumerable`1[T] manifestResources, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification peVerify, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures) [0x00055] in <2741b86d93474cfcb49ec1def5701445>:0 
          at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.Emit (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Collections.Generic.IEnumerable`1[T] manifestResources, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, System.Action`1[T] assemblyValidator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) [0x00016] in <2741b86d93474cfcb49ec1def5701445>:0 
          at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.CompileAndVerifyCommon (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] manifestResources, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Action`1[T] sourceSymbolValidator, System.Action`1[T] assemblyValidator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) [0x00074] in <2741b86d93474cfcb49ec1def5701445>:0 
          at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerify (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] manifestResources, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Action`1[T] sourceSymbolValidator, System.Action`1[T] validator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) [0x00016] in <8dbe2055f3624f378569d87082dcb029>:0 
          at Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.DefaultInterfaceImplementationTests.ValidateMethodReAbstraction_03 (System.String source1, System.String source2) [0x00062] in <6d6724c69f4c47169ca8190d23662b72>:0 
          at Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.DefaultInterfaceImplementationTests.MethodReAbstraction_03 () [0x0000d] in <6d6724c69f4c47169ca8190d23662b72>:0 
          at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
          at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <04789743042047e7ac7c5211841abf13>:0 
        ----- Inner Stack Trace -----
          at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
          at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <04789743042047e7ac7c5211841abf13>:0 
```  

40 C# compiler unit-tests fail this way for https://github.com/dotnet/roslyn/pull/35756. Complete log file is attached.
[Microsoft.CodeAnalysis.CSharp.Symbol.UnitTests_net472_x64.log](https://github.com/mono/mono/files/3188264/Microsoft.CodeAnalysis.CSharp.Symbol.UnitTests_net472_x64.log)

**Comment by marek-safar** [comment]
@lambdageek more DIM issues coming in (we'll need to backport this)

**Comment by lambdageek** [comment]
@thaystg More DIM feature work.  The CoreCLR PR has some nice tests cases.

**Comment by lambdageek** [comment]
Let's keep this open until #15229 is merged.

**Comment by thaystg** [comment]
Merged.

**Comment by AlekseyTs** [comment]
@thaystg It look like compiler tests are still failing (see https://github.com/dotnet/roslyn/pull/36712). Is this expected?

**Comment by thaystg** [comment]
> @thaystg It look like compiler tests are still failing (see [dotnet/roslyn#36712](https://github.com/dotnet/roslyn/pull/36712)). Is this expected?

I'm checking and I'll tell you soon.

**Comment by thaystg** [comment]
@AlekseyTs , which version of mono are you using? I've just tested with master and everything is working...

**Comment by AlekseyTs** [comment]
@thaystg 

> which version of mono are you using?

It looks like our CI is using 6.3.0.688-0nightly1+ubuntu1604b1

**Comment by thaystg** [comment]
@AlekseyTs https://github.com/mono/mono/commit/057f51d833f3d103730ad88a427407e3de251f6d

**Comment by AlekseyTs** [comment]
I am not sure how to map that to the package version

**Comment by lambdageek** [comment]
I asked @directhex to see why our [nightly master packages](https://download.mono-project.com/repo/debian/dists/nightly-stretch/main/binary-amd64/) are so far behind.

Background info:

6.3.0.688 is `master` after we branched `2019-04` but before we branched for `2019-06`. (688 is the number of commits since the version bump). That's a pretty old build - I'm not sure how it's being provisioned.  Currently we're building packages from master with versions like `6.5.0.nnn` ([OSX per commit packages](https://jenkins.mono-project.com/view/Per-commit%20packages/job/build-package-osx-mono/job/master/) are up to date)

More background info: [Mono versions](https://www.mono-project.com/docs/about-mono/versioning/#mono-source-versioning) - the `master versions` always have an odd number that's bumped just after we branch a `yyyy-mm` branch.
